flow/src/main/java/com/sequenceiq/flow/core/Flow2Handler.java

```diff
@@ -184,11 +193,24 @@ private void handleFlowControlEvent(String key, Payload payload, FlowParameters
         LOGGER.debug("flow control event arrived: key: {}, flowid: {}, usercrn: {}, payload: {}", key, flowId, flowParameters.getFlowTriggerUserCrn(), payload);
         Flow flow = runningFlows.get(flowId);
         if (flow != null) {
+            MutableBoolean isFlowCancelled = new MutableBoolean(false);
             transactionService.required(() -> {
                 Optional<FlowLog> lastFlowLog = flowLogService.getLastFlowLog(flow.getFlowId());
-                lastFlowLog.ifPresent(flowLog -> updateFlowLogStatus(key, payload, flowChainId, flow, flowLog, flowParameters));
+                lastFlowLog.ifPresent(flowLog -> {
+                    String nodeId = nodeConfig.getId();
+                    if (flowLog.getCloudbreakNodeId() == null || flowLog.getCloudbreakNodeId().equals(nodeId)) {
+                        updateFlowLogStatus(key, payload, flowChainId, flow, flowLog, flowParameters);
+                    } else {
+                        LOGGER.warn("Flow {} was handled by another node {}, current node ID is {}, abandoning.",
+                                flow.getFlowId(), flowLog.getCloudbreakNodeId(), nodeId);
+                        inMemoryCleanup.cancelFlowWithoutDbUpdate(flow.getFlowId());
+                        isFlowCancelled.setTrue();
+                    }
+                });
             });
-            flow.sendEvent(key, flowParameters.getFlowTriggerUserCrn(), payload, flowParameters.getSpanContext());
+            if (!isFlowCancelled.booleanValue()) {
```

Comment:
why don't send this inside the lamda?

Reply:
it may take some time to consume events, and it shouldn't be in a transaction then

Reply:
I see, thanks

Reply:
because that would change the logic: the sendEvent is necessary in the case lastFlowLog is not present